### ISSUE TITLE
LoRA gradients not normalized by input norm → training instability (NaN)

### ISSUE THREAD

**Issue by warjohn** [huggingface/peft#3073]
### System Info

Found an issue: PEFT scales LoRA output by α / r​, but gradients remain linearly dependent on input norm ∥x∥∥x∥:
∥Grad∥∝α / r⋅∥x∥ 
When activations vary 10–100× across layers, this causes:
    Gradient explosion → NaN on first steps
    Vanishing gradients in layers with small activations
Workaround: Added a hook that normalizes gradients by input energy:
scale=α / r * mean(x^T*x)​
Without this — NaN on step 1–5. With this — stable training.
Suggestion: Add normalize_grad_by_input=True option to LoraConfig or document the need for grad_norm monitoring. Something like that ->
```python
def _scale_lora_grads_by_input_norm(model, saved_inputs, lora_modules, r, alpha, eps=1e-8):
    for module in lora_modules:
        mid = id(module)
        if mid not in saved_inputs:
            continue
        x = saved_inputs[mid]
        if x is None or x.numel() == 0:
            continue
        # s = mean over batch of (x^T x) + eps
        s = (x.float() ** 2).sum(dim=-1).mean().item() + eps
        scale = alpha / (r * s)
        for lora_A in getattr(module, "lora_A", {}).values():
            if hasattr(lora_A, "weight") and lora_A.weight.grad is not None:
                lora_A.weight.grad.data.mul_(scale)
        for lora_B in getattr(module, "lora_B", {}).values():
            if hasattr(lora_B, "weight") and lora_B.weight.grad is not None:
                lora_B.weight.grad.data.mul_(scale)
```
P.s. This normalization is only needed for the first epoch, then the gradients are scaled and no spikes occur.

If my understanding is incorrect or I'm missing something, please feel free to correct me.

### Reproduction

I was working on a knowledge distillation task: distilling embeddings from a teacher model (deepvk/USER-bge-m3, 1024-dim) to a student model (ModernBERT-base, 768-dim) using LoRA adapters via PEFT.
Setup:

Student: ModernBERT + LoRA (r=16, alpha=64) on last 5 layers (later expanded to all 22 layers)
Projection head: Linear(768 → 1024) to match teacher dimension
Loss: Cosine embedding loss +
Pooling: Mean pooling (aligned between teacher and student after debugging)

What I observed:
NaN on early steps (1–5):
Training would crash with NaN loss almost immediately, especially on batches with longer sequences or high-activation samples. Gradient clipping (clip_grad_norm_=1.0) did not help.

Debugging process:
First, I checked pooling alignment: discovered I was using CLS pooling for the teacher (BGE) and mean pooling for the student. After aligning both to mean pooling, the plateau disappeared — but NaNs on early steps persisted.

Next, I inspected gradient norms per parameter group. Found that LoRA adapters had extremely high gradient norms (>1e4) compared to the projection layer, especially in deeper layers.

Instrumented forward hooks to log input activations x to each LoRA linear layer. Discovered that input norms ‖x‖ varied by 10–100× across layers and batches.

Realized that PEFT's backward pass computes gradients as ∂L/∂A ∝ (α/r) · […] · xᵀ  meaning gradient magnitude scales linearly with ‖x‖. No normalization by input energy is applied.

### Expected behavior

Hypothesis:
When input activations are large, gradients explode → NaN. When small, gradients vanish → no learning. The fixed scaling factor α/r is not enough to stabilize training across varying activation magnitudes.
Tested fix:
Added a forward hook to save inputs x, then a backward hook to rescale LoRA gradients by:
scale = α / (r · mean(xᵀx) + ε)

**Comment by githubnemo** [comment]
Thank you for your report. This is an interesting finding and props for digging in and figuring out a solution.

When it comes to what to do with this information, I'm not quite sure. The solution you found is a departure from how LoRA is described in the original paper and it would be a backwards incompatible change, so we can definitely not just add it by default. It would be possible to add this as an optional feature, if the overhead isn't too big memory-wise. However, I wonder how frequent this issue is in general.

One thing you didn't mention is if you are using any kind of gradient clipping or learning rate schedule. My suspicion is that for many users, the problem of exploding gradients is mitigated by those. This may not be the most efficient solution in terms of runtime, but it means you don't have to worry too much about exploding gradients and after a couple of steps, as you mentioned, the gradients should get into normal range. I would also recommend to check the various initialization options we have for LoRA, e.g. `"orthogonal"`.

**Comment by warjohn** [comment]
I tried several approaches to address the issue:

I used global gradient clipping with various threshold values, ranging from 0.5 down to 0.05. (As far as I know, the standard PEFT/Transformers tools don't support per-layer gradient clipping without writing custom hooks, so I stuck with global clipping. Please correct me if I'm wrong).

I experimented both with and without a learning rate scheduler. I tried the linear scheduler and the cosine scheduler.

I also tested different initialization methods, including orthogonal, and I thoroughly checked the entire dataset, suspecting the issue might lie there, but that didn't turn out to be the cause.

Could you please explain in more detail why the solution I proposed conflicts with the core logic of LoRA? I would really like to dive deeper into this and better understand the theoretical reasoning behind it.
If this is documented anywhere, could you please share a link to the paper or mathematical justification? I'd like to understand the issue from all perspectives.
Thank you in advance for your response.

**Comment by githubnemo** [comment]
> I tried several approaches to address the issue:
> 
> I used global gradient clipping with various threshold values, ranging from 0.5 down to 0.05. (As far as I know, the standard PEFT/Transformers tools don't support per-layer gradient clipping without writing custom hooks, so I stuck with global clipping. Please correct me if I'm wrong).
> 
> I experimented both with and without a learning rate scheduler. I tried the linear scheduler and the cosine scheduler.
> 
> I also tested different initialization methods, including orthogonal, and I thoroughly checked the entire dataset, suspecting the issue might lie there, but that didn't turn out to be the cause.

Thanks for the detailed description, it really helps! In your original post you mentioned `clip_grad_norm_` which is short for `torch.nn.utils.clip_grad_norm_`? Are you sure you applied the gradient norm at the correct place? When using HF trainer this is usually enforced via the `max_grad_norm` training argument. It sounds a bit strange to me that setting the grad norm as low as 0.05 did not prevent NaN values. Is there a grad norm that solves the issue?

If we want to further understanding of the issue I think we need a minimal reproducer.

> Could you please explain in more detail why the solution I proposed conflicts with the core logic of LoRA? I would really like to dive deeper into this and better understand the theoretical reasoning behind it. If this is documented anywhere, could you please share a link to the paper or mathematical justification? I'd like to understand the issue from all perspectives. Thank you in advance for your response.

What I meant was that it is a deviation of the original LoRA formulation as described in the paper and, as such, cannot be implemented as the default. But if it turns out to be an issue that needs to be resolved we can add the proposed fix as an optional feature.

**Comment by warjohn** [comment]
Sorry for keeping you waiting and taking up so much of your time. I finally figured out where the error was. The problem turned out to be that activations got out of control from the very first step. The reason was in the configuration: the combination of RSLoRA scaling and learning rate was too aggressive for this architecture. Because of this, the gradient exploded via the chain rule. I've fixed the error by adjusting the parameters, and now I'm searching for a more stable configuration for long-term training. Thanks again for your time and help!

**Comment by githubnemo** [comment]
Great that you found the reason for your issue. I'm closing this issue. Feel free to re-open if something changes.
